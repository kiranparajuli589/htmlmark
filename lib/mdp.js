import { Lexer } from "./lexer/index.js"
import { Parser } from "./parser/index.js"

/**
 * Markdown Parser for Javascript Application
 */
export class MDP {
	indent
	codeHighlightFn
	constructor(config = {}) {
		this.indent = config.indent || 4
		this.codeHighlightFn = config.codeHighlightFn || null
	}
	/**
	 * Runs the HTML parsing for the given lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	h(lines) {
		const lexer = new Lexer(lines)
		const parser = new Parser(lexer.run(), null, this.codeHighlightFn)
		return parser.run()
	}

	/**
	 * Runs HTML parsing for the lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	hT(lines) {
		const start = Date.now()
		const html = this.h(lines)
		const end = Date.now()
		console.info(`\x1b[35mParsing time: ${end - start}ms\x1b[0m`)
		return html
	}

	/**
	 * For demo preview
	 *
	 * @param {String} str - The string to parse
	 * @param {Function|null} codeHighlightFn - A function to highlight code blocks
	 *
	 * @returns {{ elapsedTime: number, html: string, lex: Object}}
	 */
	hP(str, codeHighlightFn = null) {
		const lines = str.split("\n")
		const start = Date.now()
		const lexer = new Lexer(lines)
		const lex = lexer.run()
		const parser = new Parser(lex, null, codeHighlightFn)
		const html = parser.run()
		const end = Date.now()
		return {
			elapsedTime: end - start,
			html,
			lex
		}
	}
}
