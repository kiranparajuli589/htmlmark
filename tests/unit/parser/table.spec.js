import {MDP} from "../../mdp.js"


const mdp = new MDP()


describe("Table Parsing", () => {
	describe("is not a table", () => {
		it("only header", () => {
			const html = mdp.h([
				"| column 1 | column 2 |"
			])
			expect(html).toMatchSnapshot()
		})
		it("only header and separator", () => {
			const html = mdp.h([
				"| column 1 | column 2 |",
				"|---|---|"
			])
			expect(html).toMatchSnapshot()
		})
		it("false separator", () => {
			const html = mdp.h([
				"| column 1 | column 2 |",
				"| - | - |",
				"| row 1 c1 | row 1 c2 |"
			])
			expect(html).toMatchSnapshot()
		})
		it("not consistent cell count", () => {
			const lines = [
				"| column 1 | column 2 |",
				"| --- |",
				"| row 1 c1 | row 1 c2 |"
			]
			const html = mdp.h(lines)
			expect(html).toMatchSnapshot()
		})
		it("other tokens in between 1", () => {
			const html = mdp.h([
				"| column 1 | column 2 |",
				"",
				"|---|---|",
				"| row 1 c1 | row 1 c2 |"
			])
			expect(html).toMatchSnapshot()
		})
		it("other tokens in between 2", () => {
			const html = mdp.h([
				"| column 1 | column 2 |",
				"|---|---|",
				"",
				"| row 1 c1 | row 1 c2 |"
			])
			expect(html).toMatchSnapshot()
		})
		it("other tokens in between 3", () => {
			const html = mdp.h([
				"| column 1 | column 2 |",
				"|---|---|",
				"| row 1 c1 | row 1 c2 |",
				"",
				"| row 2 c1 | row 2 c2 |"
			])
			expect(html).toMatchSnapshot()
		})
	})
	describe("table indent", () => {
		it("different indent header -> separator", () => {
			const html = mdp.h([
				"| column 1 | column 2 |",
				"    |---|---|",
				"| row 1 c1 | row 1 c2 |"
			])
			expect(html).toMatchSnapshot()
		})

		it("different indent header, separator -> body row", () => {
			const html = mdp.h([
				"    | column 1 | column 2 |",
				"    |---|---|",
				"| row 1 c1 | row 1 c2 |"
			])
			expect(html).toMatchSnapshot()
		})

		it("the indented table", () => {
			const html = mdp.h([
				"  | column 1 | column 2 |",
				"  |---|---|",
				"  | row 1 c1 | row 1 c2 |"
			])
			expect(html).toMatchSnapshot()
		})

		it("acceptable indent should not break the table 1", () => {
			const html = mdp.h([
				"  | column 1 | column 2 |",
				"  |---|---|",
				"  | row 1 c1 | row 1 c2 |",
				"| row 2 c1 | row 2 c2 |"
			])
			expect(html).toMatchSnapshot()
		})
		it("indent should break the table 2", () => {
			const html = mdp.h([
				"| column 1 | column 2 |",
				"|---|---|",
				"| row 1 c1 | row 1 c2 |",
				"    | column 1 | column 2 |",
				"    |---|---|",
				"    | row 1 c1 | row 1 c2 |"
			])
			expect(html).toMatchSnapshot()
		})
	})

	describe("table cell count", () => {
		it("cell count should break the table 1", () => {
			const html = mdp.h([
				"| column 1 | column 2 |",
				"|---|---|",
				"| row 1 c1 | row 1 c2 |",
				"| row 2 c1 |"
			])
			expect(html).toMatchSnapshot()
		})
		it("cell count should break the table 2", () => {
			const html = mdp.h([
				"| column 1 | column 2 |",
				"|---|---|",
				"| row 1 c1 | row 1 c2 |",
				"| column 1 |",
				"|---|",
				"| row 1 c1 |"
			])
			expect(html).toMatchSnapshot()
		})
	})
	describe("table heading-body separator", () => {
		it.each([
			"|-------|-------|",
			"|:-------:|:---------:|"
		])("should parse a table", (line) => {
			const html = mdp.h([
				"| column 1 | column 2 |",
				line,
				"| row 1 c1 | row 1 c2 |",
				"| row 2 c1 | row 2 c2 |"
			])
			expect(html).toMatchSnapshot()
		})
	})
	describe("pipe escaping", () => {
		it("should escape the pipe char inside table heading and or cell", () => {
			const lines = [
				"| one | two\\|three |",
				"| -- | -- |",
				"| one | two\\|five |"
			]
			const html = mdp.h(lines)
			expect(html).toMatchSnapshot()
		})
	})
})
