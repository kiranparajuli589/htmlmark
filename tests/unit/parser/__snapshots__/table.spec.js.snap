// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Parsing is not a table false separator 1`] = `
"<p>| column 1 | column 2 | | - | - | | row 1 c1 | row 1 c2 |</p>
"
`;

exports[`Table Parsing is not a table not consistent cell count 1`] = `
"<p>| column 1 | column 2 | | --- | | row 1 c1 | row 1 c2 |</p>
"
`;

exports[`Table Parsing is not a table only header 1`] = `
"<p>| column 1 | column 2 |</p>
"
`;

exports[`Table Parsing is not a table only header and separator 1`] = `
"<p>| column 1 | column 2 | |---|---|</p>
"
`;

exports[`Table Parsing is not a table other tokens in between 1 1`] = `
"<p>| column 1 | column 2 |</p>
<p>|---|---| | row 1 c1 | row 1 c2 |</p>
"
`;

exports[`Table Parsing is not a table other tokens in between 2 1`] = `
"<p>| column 1 | column 2 | |---|---|</p>
<p>| row 1 c1 | row 1 c2 |</p>
"
`;

exports[`Table Parsing is not a table other tokens in between 3 1`] = `
"<table>
  <thead>
    <tr>
      <th>column 1</th><th>column 2</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>
<p>| row 2 c1 | row 2 c2 |</p>
"
`;

exports[`Table Parsing pipe escaping should escape the pipe char inside table heading and or cell 1`] = `
"<table>
  <thead>
    <tr>
      <th>one</th><th>two|three</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>one</td><td>two|five</td></tr>
  </tbody>
</table>
"
`;

exports[`Table Parsing table cell count cell count should break the table 1 1`] = `
"<table>
  <thead>
    <tr>
      <th>column 1</th><th>column 2</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>
<p>| row 2 c1 |</p>
"
`;

exports[`Table Parsing table cell count cell count should break the table 2 1`] = `
"<table>
  <thead>
    <tr>
      <th>column 1</th><th>column 2</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th>column 1</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td></tr>
  </tbody>
</table>
"
`;

exports[`Table Parsing table heading-body separator should parse a table 1`] = `
"<table>
  <thead>
    <tr>
      <th>column 1</th><th>column 2</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr><tr><td>row 2 c1</td><td>row 2 c2</td></tr>
  </tbody>
</table>
"
`;

exports[`Table Parsing table heading-body separator should parse a table 2`] = `
"<table>
  <thead>
    <tr>
      <th>column 1</th><th>column 2</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr><tr><td>row 2 c1</td><td>row 2 c2</td></tr>
  </tbody>
</table>
"
`;

exports[`Table Parsing table indent acceptable indent should not break the table 1 1`] = `
"<table>
  <thead>
    <tr>
      <th>column 1</th><th>column 2</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>
<p>| row 2 c1 | row 2 c2 |</p>
"
`;

exports[`Table Parsing table indent different indent header -> separator 1`] = `
"<p>| column 1 | column 2 |     |---|---| | row 1 c1 | row 1 c2 |</p>
"
`;

exports[`Table Parsing table indent different indent header, separator -> body row 1`] = `
"<pre><code>| column 1 | column 2 |
|---|---|</code></pre>
<p>| row 1 c1 | row 1 c2 |</p>
"
`;

exports[`Table Parsing table indent indent should break the table 2 1`] = `
"<table>
  <thead>
    <tr>
      <th>column 1</th><th>column 2</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th>column 1</th><th>column 2</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>
"
`;

exports[`Table Parsing table indent the indented table 1`] = `
"<table>
  <thead>
    <tr>
      <th>column 1</th><th>column 2</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>
"
`;
